spotbugs/src/main/java/edu/umd/cs/findbugs/detect/CompoundOperationsOnSharedVariables.java

```diff
@@ -0,0 +1,210 @@
+import edu.umd.cs.findbugs.util.MultiThreadedCodeIdentifierUtils;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.classfile.Method;
+
+public class CompoundOperationsOnSharedVariables extends OpcodeStackDetector {
+
+    private final BugAccumulator bugAccumulator;
+    private final Map<XMethod, List<XField>> compoundlyWrittenNotSecuredFieldsByMethods;
+    private final Map<XMethod, List<XField>> readNotSecuredFieldsByMethods;
+    private boolean isInsideSynchronizedOrLockingMethod;
+    private Optional<XField> maybeCompoundlyOperatedField;
+    private int stepsMadeInCompoundOperationProcess;
+
+    public CompoundOperationsOnSharedVariables(BugReporter bugReporter) {
+        this.bugAccumulator = new BugAccumulator(bugReporter);
+        this.compoundlyWrittenNotSecuredFieldsByMethods = new HashMap();
+        this.readNotSecuredFieldsByMethods = new HashMap();
+        this.isInsideSynchronizedOrLockingMethod = false;
+        this.maybeCompoundlyOperatedField = Optional.empty();
+        this.stepsMadeInCompoundOperationProcess = Const.UNDEFINED;
+    }
+
+    @Override
+    public void visit(Method method) {
+        ClassContext currentClassContext = getClassContext();
+        isInsideSynchronizedOrLockingMethod = MultiThreadedCodeIdentifierUtils.isMethodMultiThreaded(method, currentClassContext);
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        bugAccumulator.reportAccumulatedBugs();
+    }
+
+    @Override
+    public void visitClassContext(ClassContext classContext) {
+        if (MultiThreadedCodeIdentifierUtils.isPartOfMultiThreadedCode(classContext)) {
+            isInsideSynchronizedOrLockingMethod = false;
+            maybeCompoundlyOperatedField = Optional.empty();
+            stepsMadeInCompoundOperationProcess = Const.UNDEFINED;
+
+            super.visitClassContext(classContext);
+        }
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (!isInsideSynchronizedOrLockingMethod) {
+            if (seen == Const.GETFIELD || seen == Const.GETSTATIC) {
+                XMethod readMethod = getXMethod();
+                Optional<XField> maybeFieldToRead = Optional.ofNullable(getXFieldOperand());
+                lookForUnsecuredOperationsOnFieldInOtherMethods(
+                        maybeFieldToRead, readMethod, compoundlyWrittenNotSecuredFieldsByMethods, readNotSecuredFieldsByMethods);
+                maybeCompoundlyOperatedField = maybeFieldToRead;
+                stepsMadeInCompoundOperationProcess = 1;
+                return;
+            }
+
+            // Possible invoke virtual first if it is a wrapper type
+            if (stepsMadeInCompoundOperationProcess == 1 && seen == Const.INVOKEVIRTUAL) {
```

Comment:
These `if...return` blocks are quite confusing at first glance. Also, some of them could be combined, eg there are two blocks that each check if stepsMadeInCompoundOperationProcess == 1. Perhaps a `switch` statement could improve the structure?